### PR TITLE
mysql:add type json

### PR DESCRIPTION
**Pull request by AndreMouche**
Hi,
     This PR add type `json` in package `mysql`, the related PR is https://github.com/pingcap/tidb/pull/3248/files

@hhkbp2 @BusyJay PTAL

### CODE REVIEW

File: src/util/codec/mysql/json.rs [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::string::String;
```
Comment (andelf) [inline src/util/codec/mysql/json.rs]: needless

File: src/util/codec/mysql/json.rs [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+const LENGTH_TYPE: usize = 1;
+const LENGTH_LITERAL: usize = 1;
+const LENGTH_U16: usize = 2;
+const LENGTH_U32: usize = 4;
+const LENGTH_KEY_ENTRY: usize = LENGTH_U32 + LENGTH_U16;
+const LENGTH_VALUE_ENTRY: usize = LENGTH_TYPE + LENGTH_U32;
+
+
+macro_rules! take_prefix {
+   ($data:ident,$len:ident) => {{
```
Comment (andelf) [inline src/util/codec/mysql/json.rs]: `$len` should be expr

File: src/util/codec/mysql/json.rs [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+        }
+    }
+}
+
+fn decode_literal(mut data: &[u8]) -> Result<JSON> {
+    let literal = take_prefix!(data, LENGTH_LITERAL);
+    Ok(JSON::JLiteral(literal[0]))
+}
+
+fn encode_str(data: String) -> Vec<u8> {
```
Comment (andelf) [inline src/util/codec/mysql/json.rs]: Better pass `buf: &mut Vec<u8>` or `mut buf: Write` as second parameter,
to avoid Vec allocation

**Comment by AndreMouche** [comment]
@hicqu PTAL

File: src/util/codec/mysql/json.rs [hunk 4]
```diff
@@ -0,0 +1,545 @@
...
+        let mut buffer = vec![0;$len as usize];
+        $data.read_exact(&mut buffer).unwrap();
+        buffer
+        }
+   }
+}
+
+#[derive(Clone,PartialEq,Debug)]
+pub enum JSON {
+    JLiteral(u8),
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: Prefix `J` is redundant.

File: src/util/codec/mysql/json.rs [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+
+const LENGTH_TYPE: usize = 1;
+const LENGTH_LITERAL: usize = 1;
+const LENGTH_U16: usize = 2;
+const LENGTH_U32: usize = 4;
+const LENGTH_KEY_ENTRY: usize = LENGTH_U32 + LENGTH_U16;
+const LENGTH_VALUE_ENTRY: usize = LENGTH_TYPE + LENGTH_U32;
+
+macro_rules! take_prefix {
+   ($data:ident,$len:expr) => {{
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: Missing space.

File: src/util/codec/mysql/json.rs [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+        let mut buffer = vec![0;$len as usize];
+        $data.read_exact(&mut buffer).unwrap();
+        buffer
+        }
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: Strange bracket.

File: src/util/codec/mysql/json.rs [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+   }
+}
+
+#[derive(Clone,PartialEq,Debug)]
+pub enum JSON {
+    JLiteral(u8),
+    JString(String),
+    JObject(BTreeMap<String, JSON>),
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: Why `BTreeMap`?

File: src/util/codec/mysql/json.rs [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+}
+
+#[allow(dead_code)]
+impl JSON {
+    pub fn deserialize(mut data: &[u8]) -> Result<JSON> {
+        if data.is_empty() {
+            return Err(invalid_type!("The document is empty"));
+        }
+        let code_type = take_prefix!(data, LENGTH_TYPE);
+        JSON::decode(code_type[0], data)
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: I think `JSON::decode(data[0], data)` is enough.

File: src/util/codec/mysql/json.rs [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+   ($data:ident,$len:expr) => {{
+        let mut buffer = vec![0;$len as usize];
+        $data.read_exact(&mut buffer).unwrap();
+        buffer
+        }
+   }
+}
+
+#[derive(Clone,PartialEq,Debug)]
+pub enum JSON {
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: Should be `Json`.

File: src/util/codec/mysql/json.rs [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+            JSON::JString(_) => TYPE_CODE_STRING,
+            JSON::JObject(_) => TYPE_CODE_OBJECT,
+            JSON::JArray(_) => TYPE_CODE_ARRAY,
+            JSON::JDouble(_) => TYPE_CODE_DOUBLE,
+            JSON::JI64(_) => TYPE_CODE_I64,
+        }
+    }
+
+    pub fn to_string(&self) -> String {
+        if let JSON::JLiteral(ref d) = *self {
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: Is the check necessary?

File: src/util/codec/mysql/json.rs [hunk 11]
```diff
@@ -0,0 +1,545 @@
...
+const JSON_LITERAL_FALSE: u8 = 0x02;
+
+const LENGTH_TYPE: usize = 1;
+const LENGTH_LITERAL: usize = 1;
+const LENGTH_U16: usize = 2;
+const LENGTH_U32: usize = 4;
+const LENGTH_KEY_ENTRY: usize = LENGTH_U32 + LENGTH_U16;
+const LENGTH_VALUE_ENTRY: usize = LENGTH_TYPE + LENGTH_U32;
+
+macro_rules! take_prefix {
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: I don't think the macro is necessary. In most case, use `&data[..len]` is enough.

File: src/util/codec/mysql/json.rs (hunk 9 above)
Comment (hicqu) [inline src/util/codec/mysql/json.rs]: I think JSON is better. In MySQL 5.7 source code, all are JSON but not Json.

File: src/util/codec/mysql/json.rs (hunk 7 above)
Comment (hicqu) [inline src/util/codec/mysql/json.rs]: @BusyJay here I think BTreeMap is better than HashMap because when encoding JSON as binary, we should keep all keys sorted.

File: src/util/codec/mysql/json.rs (hunk 9 above)
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: No, in rust, acronyms count as one word.

Comment (AndreMouche) [inline src/util/codec/mysql/json.rs]: I'd like to use `Json` here according to the [ Rust's Naming conventions](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html ) @hicqu

File: src/util/codec/mysql/json.rs [hunk 12]
```diff
@@ -0,0 +1,472 @@
...
+    // element-count ::= uint32 number of members in object or number of elements in array
+    let element_count = data.len();
+    let element_count_len = LENGTH_U32;
+    // size ::= uint32 number of bytes in the binary representation of the object or array
+    let size_len = LENGTH_U32;
+    // key-entry ::= key-offset(uint32) key-length(uint16)
+    let key_entries_len = LENGTH_KEY_ENTRY * element_count;
+    // value-entry ::= type(byte) offset-or-inlined-value(uint32)
+    let value_entries_len = LENGTH_VALUE_ENTRY * element_count;
+    let mut keys = Vec::with_capacity(element_count);
```
Comment (hicqu) [inline src/util/codec/mysql/json.rs]: maybe we dont't need realloc these two Vector of string `keys` and `values`? 
I think we can pass an iterator to `encode_vec_json`:
```rust
let values = data.iter().map(|(k, v)| v);
encode_vec_json(values, offset);
func encode_vec_json(it: &mut Iterator<Item=&str>, offset: u32) {
    // ...
}
```
what do you think?

Comment (AndreMouche) [inline src/util/codec/mysql/json.rs]: Since we couldn't independently take ownership of the value  from a map. It seems not a good idea since I would like to `take`  the `key,value`  from the `data` instead of `clone`.  @hicqu

File: src/util/codec/mysql/json.rs [hunk 13]
```diff
@@ -0,0 +1,472 @@
...
+    Ok(Json::Literal(data[0]))
+}
+
+fn encode_str(data: String, buf: &mut Vec<u8>) {
+    buf.encode_var_u64(data.len() as u64).unwrap();
+    buf.append(&mut data.into_bytes());
+}
+
+fn decode_str(mut data: &[u8]) -> Result<Json> {
+    let length = try!(data.decode_var_u64());
```
Comment (siddontang) [inline src/util/codec/mysql/json.rs]: `decode_var_u64` is the same with https://github.com/mysql/mysql-server/blob/23032807537d8dd8ee4ec1c4d40f0633cd4e12f9/sql/json_binary.cc#L234 ?

Comment (AndreMouche) [inline src/util/codec/mysql/json.rs]: yes. Since we definitely want to read a `u64`,  so we needn't to pass the number to read as parameter.  Similarly, we needn't return the length we read. @siddontang

**Comment by AndreMouche** [comment]
@hicqu @tiancaiamao PTAL

File: src/util/codec/number.rs [hunk 14]
```diff
@@ -110,6 +110,22 @@ pub trait NumberEncoder: Write {
         let u = order_encode_f64(f);
         self.encode_u64_desc(u)
     }
+
+    fn encode_u16_with_little_endian(&mut self, v: u16) -> Result<()> {
```
Comment (BusyJay) [inline src/util/codec/number.rs]: `encode_u16_le` should be enough.

File: src/util/codec/number.rs [hunk 15]
```diff
@@ -346,9 +443,6 @@ mod test {
     test_eof!(i64_eof, encode_i64, decode_i64, 1);
     test_eof!(u64_eof, encode_u64, decode_u64, 1);
     test_eof!(f64_eof, encode_f64, decode_f64, 1.0);
-    test_eof!(i64_desc_eof, encode_i64_desc, decode_i64_desc, 1);
```
Comment (BusyJay) [inline src/util/codec/number.rs]: Why remove these?

File: src/util/codec/mysql/json.rs [hunk 16]
```diff
@@ -0,0 +1,454 @@
...
+impl Json {
+    pub fn deserialize(data: &[u8]) -> Result<Json> {
+        if data.is_empty() {
+            return Err(invalid_type!(ERR_EMPTY_DOCUMENT));
+        }
+        Json::decode(data[0], &data[LENGTH_TYPE..])
+    }
+
+    pub fn parse_from_string(s: &str) -> Result<Json> {
+        if s.is_empty() {
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: Does the check necessary?

File: src/util/codec/mysql/json.rs [hunk 17]
```diff
@@ -0,0 +1,454 @@
...
+#[allow(dead_code)]
+impl Json {
+    pub fn deserialize(data: &[u8]) -> Result<Json> {
+        if data.is_empty() {
+            return Err(invalid_type!(ERR_EMPTY_DOCUMENT));
+        }
+        Json::decode(data[0], &data[LENGTH_TYPE..])
+    }
+
+    pub fn parse_from_string(s: &str) -> Result<Json> {
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: s/parse_from_string/from_str/

File: src/util/codec/mysql/json.rs [hunk 18]
```diff
@@ -0,0 +1,454 @@
...
+        if s.is_empty() {
+            return Err(invalid_type!(ERR_EMPTY_DOCUMENT));
+        }
+        match serde_json::from_str(s) {
+            Ok(value) => Ok(normalize(value)),
+            Err(e) => Err(invalid_type!("Illegal Json text:{:?}", e)),
+        }
+    }
+
+    pub fn serialize(&self) -> Vec<u8> {
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: You should define an Encoder just like DecimalEncoder.

File: src/util/codec/mysql/json.rs [hunk 19]
```diff
@@ -0,0 +1,454 @@
...
+    // element-count ::= uint32 number of members in object or number of elements in array
+    let element_count = data.len();
+    let element_count_len = LENGTH_U32;
+    // size ::= uint32 number of bytes in the binary representation of the object or array
+    let size_len = LENGTH_U32;
+    // key-entry ::= key-offset(uint32) key-length(uint16)
+    let key_entries_len = LENGTH_KEY_ENTRY * element_count;
+    // value-entry ::= type(byte) offset-or-inlined-value(uint32)
+    let value_entries_len = LENGTH_VALUE_ENTRY * element_count;
+    let mut key_entries = vec![];
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: Use `with_capacity` instead.

File: src/util/codec/mysql/json.rs (hunk 17 above)
Comment (hicqu) [inline src/util/codec/mysql/json.rs]: we have two way to create a Json from string:
1) using that string as a primitive value,  e.g. `"hello", "1", ...`
2) parse from that string, e.g. `"\"hello\"", "[1, 2, 3]", ...`

we can implement (1) by `impl From<String> for Json`, so  we get `Json::from("abcde)`.
But which style is best for (2)? `from_str` or `parse_XXX`?

Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: `FromStr` is a generic trait for string parsing. I don't think it's necessary to implement From<String> for Json. `Json::from` doesn't look better than `Json::String`.

**Comment by hicqu** [comment]
LGTM.

**Comment by AndreMouche** [comment]
@BusyJay @andelf @siddontang PTAL, THX

File: src/util/codec/mysql/json.rs [hunk 20]
```diff
@@ -0,0 +1,515 @@
...
+                let mut value = &values_data[offset_in_values as usize..];
+                value.decode_json_body(code)
+            }
+        }
+    }
+}
+
+impl<T: Read> JsonDecoder for T {}
+
+fn normalize(data: Value) -> Json {
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: Implement `Deserialize` instead.

Comment (AndreMouche) [inline src/util/codec/mysql/json.rs]: It seems may make the code structure more complicated since we should not only implement [Deserialize](https://serde.rs/impl-deserialize.html), but also need to implement  [Visiter](https://docs.serde.rs/serde/de/trait.Visitor.html) @BusyJay ?

Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: Yes. But in that way it will only iterate json once and allocate once.

File: src/util/codec/mysql/json.rs [hunk 21]
```diff
@@ -0,0 +1,562 @@
...
+        Ok(Json::Object(map))
+    }
+}
+
+impl<'de> Deserialize<'de> for Json {
+    fn deserialize<D>(deserializer: D) -> SResult<Self, D::Error>
+        where D: Deserializer<'de>
+    {
+        let visitor = JsonVisitor {};
+        deserializer.deserialize_any(visitor)
```
Comment (andelf) [inline src/util/codec/mysql/json.rs]: deserializer.deserialize_any(JsonVisitor)

File: src/util/codec/number.rs [hunk 22]
```diff
@@ -172,6 +188,22 @@ pub trait NumberDecoder: Read {
     fn decode_f64_desc(&mut self) -> Result<f64> {
         self.decode_u64_desc().map(order_decode_f64)
     }
+
+    fn decode_u16_le(&mut self) -> Result<u16> {
+        self.read_u16::<LittleEndian>().map_err(From::from)
+    }
+
+    fn decode_u32_le(&mut self) -> Result<u32> {
+        self.read_u32::<LittleEndian>().map_err(From::from)
```
Comment (andelf) [inline src/util/codec/number.rs]: try! instead of From::from

**Comment by andelf** [comment]
LGTM

File: src/util/codec/mysql/json.rs [hunk 23]
```diff
@@ -0,0 +1,561 @@
...
+        match *self {
+            Json::Object(ref d) => get_obj_binary_len(d),
+            Json::Array(ref d) => get_array_binary_len(d),
+            Json::Literal(_) => LENGTH_LITERAL,
+            Json::I64(_) | Json::Double(_) => LENGTH_I64,
+            Json::String(ref d) => get_str_binary_len(d),
+        }
+    }
+
+    pub fn to_string(&self) -> String {
```
Comment (tennix) [inline src/util/codec/mysql/json.rs]: return `Result<String>`

File: src/util/codec/mysql/json.rs [hunk 24]
```diff
@@ -0,0 +1,561 @@
...
+        where E: de::Error
+    {
+        if v {
+            Ok(Json::Literal(JSON_LITERAL_TRUE))
+        } else {
+            Ok(Json::Literal(JSON_LITERAL_FALSE))
+        }
+    }
+
+    fn visit_i64<E>(self, v: i64) -> SResult<Self::Value, E>
```
Comment (tennix) [inline src/util/codec/mysql/json.rs]: Is it necessary to implement visit_i8, visit_i16, visit_f32 ... methods?

File: src/util/codec/number.rs [hunk 25]
```diff
@@ -110,6 +110,22 @@ pub trait NumberEncoder: Write {
         let u = order_encode_f64(f);
         self.encode_u64_desc(u)
     }
+
+    fn encode_u16_le(&mut self, v: u16) -> Result<()> {
+        self.write_u16::<LittleEndian>(v).map_err(From::from)
```
Comment (tennix) [inline src/util/codec/number.rs]: You can use `try!` instead of `map_err(From::from)` here

Comment (BusyJay) [inline src/util/codec/number.rs]: I prefer `map_err` here, because it fits in one line.

Comment (andelf) [inline src/util/codec/number.rs]: This is for consistency with above all.

File: src/util/codec/number.rs [hunk 26]
```diff
@@ -273,7 +345,7 @@ mod test {
             #[allow(unused_imports)]
             #[allow(float_cmp)]
             mod $enc {
-                use super::{I64_TESTS, U64_TESTS, F64_TESTS};
+                use super::{I64_TESTS, U64_TESTS, U32_TESTS,U16_TESTS,F64_TESTS};
```
Comment (andelf) [inline src/util/codec/number.rs]: space after comma

File: src/util/codec/number.rs (hunk 25 above)
Comment (BusyJay) [inline src/util/codec/number.rs]: There are many places that use `map_err`, using `try!` or `?` doesn't mean map_err should be forbidden.

File: src/util/codec/mysql/json.rs [hunk 27]
```diff
@@ -0,0 +1,561 @@
...
+            Json::I64(d) => serializer.serialize_i64(d),
+        }
+    }
+}
+
+struct JsonVisitor;
+impl<'de> Visitor<'de> for JsonVisitor {
+    type Value = Json;
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "may be any string")
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: What does 'may be any string' mean?

File: src/util/codec/mysql/json.rs [hunk 28]
```diff
@@ -0,0 +1,561 @@
...
+                try!(self.write_u8(*v));
+                let left = LENGTH_U32 - LENGTH_LITERAL;
+                for _ in 0..left {
+                    try!(self.write_u8(JSON_LITERAL_NIL));
+                }
+            }
+            _ => {
+                try!(self.encode_u32_le(*offset));
+                try!(data_buf.encode_json_body(data));
+                *offset += (data.binary_len() - LENGTH_TYPE) as u32;
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: I think `data.binary_len() - LENGTH_TYPE` should be equal to `length of data_buf - original length of data_buf`.

File: src/util/codec/mysql/json.rs [hunk 29]
```diff
@@ -0,0 +1,561 @@
...
+        let count_len = LENGTH_U32;
+        let size_len = LENGTH_U32;
+        let value_entries_len = LENGTH_VALUE_ENTRY * element_count;
+        let mut value_offset = (count_len + size_len + value_entries_len) as u32;
+        let mut value_entries = Vec::with_capacity(value_entries_len);
+        let mut encode_values = vec![];
+        for value in data {
+            try!(value_entries.encode_json_item(value, &mut value_offset, &mut encode_values));
+        }
+        let total_size = value_offset;
```
Comment (BusyJay) [inline src/util/codec/mysql/json.rs]: I think `total_size` should be `encode_values.len() + count_len + size_len + value_entries_len`.

File: src/util/codec/mysql/json.rs (hunk 24 above)
Comment (AndreMouche) [inline src/util/codec/mysql/json.rs]: It's not necessary to implement since [`visit_i8`](https://docs.serde.rs/src/serde/de/mod.rs.html#1091), `visit_i16`'s implementations are based on `visit_i64`, so does `visit_f32` @tennix

Comment (tennix) [inline src/util/codec/mysql/json.rs]: ok, fair enough

File: src/util/codec/mysql/json.rs (hunk 27 above)
Comment (AndreMouche) [inline src/util/codec/mysql/json.rs]: `expecting` is used  in error messages, and since the string we expect may be one of `string`,`int`,`array`, `object`,`boolean` and `null`, so I use  `"may be any string"` here. 
https://docs.serde.rs/serde/de/trait.Visitor.html#tymethod.expecting @BusyJay

**Comment by AndreMouche** [comment]
@andelf @hicqu PTAL agin,THX

**Comment by AndreMouche** [comment]
@hicqu PTAL again, thx

**Comment by hicqu** [comment]
LGTM.

File: src/util/codec/mysql/json.rs [hunk 30]
```diff
@@ -0,0 +1,566 @@
...
+
+const LENGTH_TYPE: usize = 1;
+const LENGTH_LITERAL: usize = 1;
+const LENGTH_U16: usize = 2;
+const LENGTH_U32: usize = 4;
+const LENGTH_I64: usize = 8;
+const LENGTH_KEY_ENTRY: usize = LENGTH_U32 + LENGTH_U16;
+const LENGTH_VALUE_ENTRY: usize = LENGTH_TYPE + LENGTH_U32;
+
+// Json implement type json used in tikv, it specifies the following
```
Comment (hhkbp2) [inline src/util/codec/mysql/json.rs]: s/implement/implements/

File: src/util/codec/mysql/json.rs [hunk 31]
```diff
@@ -0,0 +1,566 @@
...
+        serde_json::to_string(self).unwrap()
+    }
+}
+
+impl FromStr for Json {
+    type Err = Error;
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        match serde_json::from_str(s) {
+            Ok(value) => Ok(value),
+            Err(e) => Err(invalid_type!("Illegal Json text:{:?}", e)),
```
Comment (hhkbp2) [inline src/util/codec/mysql/json.rs]: s/text:{/test: {/

File: src/util/codec/mysql/json.rs [hunk 32]
```diff
@@ -0,0 +1,566 @@
...
+    let element_count = data.len();
+    let value_entries_len = element_count * LENGTH_VALUE_ENTRY;
+    let mut values_len = 0;
+    for v in data {
+        if v.get_type_code() != TYPE_CODE_LITERAL {
+            values_len += v.binary_len() - LENGTH_TYPE;
+        }
+    }
+    // array ::= element-count size value-entry* value*
+    LENGTH_U32 + LENGTH_U32 + value_entries_len + values_len
```
Comment (hhkbp2) [inline src/util/codec/mysql/json.rs]: Define ELEMENT_COUNT_LEN and SIZE_LEN for better understanding.

File: src/util/codec/mysql/json.rs [hunk 33]
```diff
@@ -0,0 +1,566 @@
...
+}
+
+#[allow(dead_code)]
+fn get_str_binary_len(data: &str) -> usize {
+    let len = data.as_bytes().len();
+    get_var_u64_blen(len as u64) + len
+}
+
+#[allow(dead_code)]
+fn get_var_u64_blen(mut v: u64) -> usize {
```
Comment (hhkbp2) [inline src/util/codec/mysql/json.rs]: What is the meaning of `blen`?

File: src/util/codec/mysql/json.rs [hunk 34]
```diff
@@ -0,0 +1,566 @@
...
+            seqs.push(value);
+        }
+        Ok(Json::Array(seqs))
+    }
+
+    fn visit_map<M>(self, mut access: M) -> std::result::Result<Self::Value, M::Error>
+        where M: MapAccess<'de>
+    {
+        let mut map = BTreeMap::new();
+
```
Comment (hhkbp2) [inline src/util/codec/mysql/json.rs]: Remove this empty line.

File: src/util/codec/mysql/json.rs [hunk 35]
```diff
@@ -0,0 +1,566 @@
...
+            Json::Literal(d) => self.encode_literal(d),
+            Json::I64(d) => self.encode_json_i64(d),
+            Json::Double(d) => self.encode_json_f64(d),
+            Json::String(ref d) => self.encode_str(d),
+        }
+    }
+
+    fn encode_obj(&mut self, data: &BTreeMap<String, Json>) -> Result<()> {
+        // object: element-count size key-entry* value-entry* key* value*
+        // element-count ::= uint32 number of members in object or number of elements in array
```
Comment (hhkbp2) [inline src/util/codec/mysql/json.rs]: Separate the comment and BNF for easier understanding.

File: src/util/codec/mysql/json.rs [hunk 36]
```diff
@@ -0,0 +1,566 @@
...
+        try!(self.write_all(value_entries.as_mut()));
+        try!(self.write_all(encode_keys.as_mut()));
+        try!(self.write_all(encode_values.as_mut()));
+        Ok(())
+    }
+
+    fn encode_array(&mut self, data: &[Json]) -> Result<()> {
+        // array ::= element-count size value-entry* value*
+        let element_count = data.len();
+        let count_len = LENGTH_U32;
```
Comment (hhkbp2) [inline src/util/codec/mysql/json.rs]: s/count_len/element_count_len/

File: src/util/codec/mysql/json.rs [hunk 37]
```diff
@@ -0,0 +1,566 @@
...
+        if element_count == 0 {
+            return Ok(Json::Object(obj));
+        }
+        // key_entries
+        let key_entries_len = LENGTH_KEY_ENTRY * element_count;
+        let mut key_entries_data = &data[0..key_entries_len];
+
+        // value-entry ::= type(byte) offset-or-inlined-value(uint32)
+        let value_entries_len = LENGTH_VALUE_ENTRY * element_count;
+        let mut value_entries_data = &data[key_entries_len..(key_entries_len + value_entries_len)];
```
Comment (hhkbp2) [inline src/util/codec/mysql/json.rs]: s/key_entries_len + value_entries_len/key_offset/

File: src/util/codec/mysql/json.rs [hunk 38]
```diff
@@ -0,0 +1,566 @@
...
+        let mut key_offset = key_entries_len + value_entries_len;
+        for _ in 0..element_count {
+            let key_real_offset = try!(key_entries_data.decode_u32_le());
+            let key_len = try!(key_entries_data.decode_u16_le());
+            let key_data = &data[key_offset..(key_offset + key_len as usize)];
+            let key = String::from(str::from_utf8(key_data).unwrap());
```
Comment (hhkbp2) [inline src/util/codec/mysql/json.rs]: s/unwrap()/try!(from_utf8())/
So that the data corruption would not lead to panic.

File: src/util/codec/mysql/json.rs [hunk 39]
```diff
@@ -0,0 +1,566 @@
...
+impl<T: Read> JsonDecoder for T {}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_json_serialize() {
+        let jstr1 =
+            r#"{"aaaaaaaaaaa": [1, "2", {"aa": "bb"}, 4.0], "bbbbbbbbbb": true, "ccccccccc": "d"}"#;
```
Comment (hhkbp2) [inline src/util/codec/mysql/json.rs]: No `null`?

File: tests/raft/test_raft.rs [hunk 40]
```diff
@@ -2633,7 +2633,7 @@ fn test_remove_node() {
 
     // remove all nodes from cluster
     r.remove_node(1);
-    assert_eq!(r.nodes(), vec![]);
+    assert!(r.nodes().is_empty());
```
Comment (hhkbp2) [inline tests/raft/test_raft.rs]: What adding this to this PR? It seems unrelated.
